duck/call.py

```diff
@@ -0,0 +1,171 @@
+
+        # Determine if the given argument(s) are present in the Call.
+        args_results = []
+        for idx, actual in enumerate(args):
+            if actual == self.args[idx]:
+                args_results.append(True)
+            else:
+                args_results.append(False)
+
+        results.append(all(args_results))
+
+        # If the given keyword arguments exceed the number of actual keyword
+        # arguments, this cannot be true either.
+        if len(kwargs) > len(self.kwargs):
+            return False
+
+        # Determine if the given keyword argument(s) are present in the Call.
+        kwargs_results = []
+        for k, v in kwargs.items():
+            if k in self.kwargs and self.kwargs[k] == v:
+                kwargs_results.append(True)
+            else:
+                kwargs_results.append(False)
+
+        results.append(all(kwargs_results))
+
+        return all(results)
+
+    def has_exact_args(self, *args, **kwargs):
+        """Returns ``True`` if the call was made with exactly the provided
+        arguments and keyword arguments, and ``False`` otherwise.
+
+        If provided, positional arguments must start at ``0`` and be in order.
+
+        Comparisons are performed with ``==``, not ``is``.
+        """
+        # There must be the same number of arguments.
+        if len(args) != len(self.args):
+            return False
+
+        # Arguments must be identical.
+        for idx, actual in enumerate(args):
+            if actual != self.args[idx]:
+                return False
+
+        # There must be the same number of keyword arguments.
+        if len(kwargs) != len(self.kwargs):
+            return False
+
+        # They must be identical.
+        for k, v in kwargs.items():
+            if k not in self.kwargs or self.kwargs[k] != v:
+                return False
+
+        return True
+
+    def assert_has_args(self, *args, **kwargs):
+        """Calls the ``has_args`` method, and raises ``AssertionError``
+        if ``False`` is returned.
+        """
+        errors = []
+
+        # Determine if the given arguments are valid.
+        if not self.has_args(*args):
+            errors.append('Given args: {} != Valid args: {}'.format(
+                args,
+                self.args,
+            ))
+
+        # Determine if the given keyword arguments are valid.
+        if not self.has_args(**kwargs):
+            errors.append('Given kwargs: {} != Valid kwargs: {}'.format(
+                kwargs,
+                self.kwargs,
+            ))
+
+        if errors:
+            raise AssertionError('\r\n'.join(errors))
+
+        return True
+
+    def assert_has_exact_args(self, *args, **kwargs):
+        """Calls the ``has_exact_args`` method, and raises ``AssertionError``
+        if ``False`` is returned.
+
+        In order to return useful debugging information, it post-processes
+        similar to ``has_exact_args``, but with additional information.
```

Comment:
I had considered something like a `self.VERBOSE` value to do this, but then we're keeping state on the object between calls, and have to deal with setting it before and resetting it back after each call. Seemed less than ideal. 

Maybe a `try/finally` even though there's no real exception handling going on?

_Sent from my Google Pixel XL using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_